Split out AST and Parser libraries from GHC (#56)
haskellfoundation/tech-proposals: proposals/accepted/000-ast-parser-libs.rst

===========================================
Split out AST and Parser libraries from GHC
===========================================

:Date: August 2023
:Authors:
  John Ericson,
  Shayne Fletcher,
  Laurent P. René de Cotret

.. sectnum::
.. contents::

Abstract
========

Problem
-------

The community lacks AST and Parser libraries for Haskell that are both self-contained and up-to-date.
Experience has shown that there is only way one way to meet each criterion:

- Be used by GHC, so the library cannot fall behind new language development

- Be separate from GHC, so the library is forced to be self-contained

However, no library has so far done both, nor met meet both criteria.

Solution
--------

The purpose of this proposal is to make that library finally exist.
The Haskell Foundation will finance the completion of the existing "Trees that grow" project, decoupling GHC's AST and parser from the rest of the compiler so they can be moved to separate libraries.

> Those libaries will be "normal" haskell libraries, without any weird dependencies or build process, and published on Hackage.

Reviewer: ```suggestion
Those libraries will be "normal" Haskell libraries, without any unusual dependencies or build processes, and published on Hackage.
```

It presents these data types:

.. code-block:: haskell

  data Component = Compiler Pass | TemplateHaskell

  data Pass = Parser | Renamer | TypeChecker

The idea that they are "promoted" via ``DataKinds``, and then type families used in the AST will have instances for these promoted values.
This allows those consumers to "adjust" the AST for their purpose.

The Trees That Grow project is now 6 years old, and has met great success in avoiding partiality in the compiler, "making illegal states unrepresentable" as many Haskellers would put it.
But progress on `reducing AST & parser dependencies <https://gitlab.haskell.org/ghc/ghc/-/issues/19932>`_ has been less easily forthcoming.
We have separated out the modules defining the AST under ``Language.Haskell.Syntax.*`` we wish to split out, and we have tests to track progress reducing their deps, and the parser's deps.
But progress is unsteady and unpredictable.

The basic problem is that the benefits don't actually kick in until the deps are *all* gone, and the code is actually separated out.

> Partial progress isn't really directly useful to anyone, and these counters just scoreboard by which we hope to get closer to the end goal.

Reviewer: What does "these counters just scoreboard by which" mean?

Author: in this paragraph I am trying to contrast the smooth function of module count hopefully going down with the step function of "benefits only kick in when count hits 0". It's good to track getting closer to 0, but important to remember partial progress != partial benefit.

Perhaps this whole paragraph is confusing and should be reworded?

Reviewer: I think what threw me was the lack of the words "are" and "a" surrounding "just" :-)

```suggestion
Partial progress isn't really directly useful to anyone, and these counters are just a scoreboard by which we hope to get closer to the end goal.
```

It is thus hard to do this work with volunteers only, because it is emphatically *not* `"itch scratching" <https://en.wikipedia.org/wiki/The_Cathedral_and_the_Bazaar>`_ work where incremental progress leads immediate incremental benefits to the contributor.

The Haskell Foundation's support in getting this "over the finish line", at which point the community *will* benefit, and benefit greatly, is thus a crucial way we can surmount the coordination failure the lack of incremental payoff causes.

.. [#intra]
  It might sound like the goal is only different usages within GHC, but remember that ``template-haskell`` is a separate library used by users of Haskell not just developers of Haskell.
  A goal of at least some usage outside GHC was always there.

Reinstallable GHC Lib
---------------------

> One of the problems ``ghc-lib-parser`` aims to solve is that ``ghc`` the library is current cumbersome to install as a regular haskell library (as opposed to by switching toolchains entirely).

Reviewer: ```suggestion
One of the problems ``ghc-lib-parser`` aims to solve is that ``ghc`` the library is current cumbersome to install as a regular Haskell library (as opposed to by switching toolchains entirely).
```

There is currently work in flight to solve that.
One that is done, projects like HLint_ *could* just depend on ``ghc`` directly, and still be easily buildable (with Cabal / with Stack / from Hackage) as today.

Just doing this isn't a good solution though, because ``ghc`` exposes a much a wider surface area than what these projects actually want.
For stability's sake, it is better that those libraries dependent on narrower parsing / AST libraries that only provide what they need.
And longer term, we hope the "tug of war" of between GHC and these projects as consumers of those libraries, versus just the others having to deal with whatever GHC does with just itself in mind, will result in a higher-quality, more flexible, and overall friendlier library.

In `this comment <https://gitlab.haskell.org/ghc/ghc/-/issues/14409#note_506489>`_, it is suggested that factoring out the AST and parser can be a good first step making a more modular in GHC in general.

> This proposal wish to *stay neutral* on the merits of such a future direction, but it would be remiss not to at least highlight it as one possible outcome.

Reviewer: ```suggestion
This proposal wishes to *stay neutral* on the merits of such a future direction, but it would be remiss not to at least highlight it as one possible outcome.
```

The timeline for this is pretty short because there exists an easy last-resort way to decouple anything:
just add another TTG type family.
This came up with some acrimony in `GHC Issue #21628 <https://gitlab.haskell.org/ghc/ghc/-/issues/21628>`_, discussing whether it was better to try to change GHC's ``FastString`` or abstract over it.
The purpose of this proposal isn't to relitigate that issue, but because this proposal *is* about resource allocation, something does need to be said on the broader tradeoffs at play

There is no disagreement that as-is, that data type is not suitable for a nice self-contained library. [#faststring-unsuitable]_
The disagreement is whether TTG should be blocked on reworking ``FastString`` somehow to be better for GHC and non-GHC alike, or whether we should just side-step the issue entirely.

We make no claims about what is better in the long term for GHC, but when reworking ``FastString`` and benchmarking the new algorthms might take **Days to Weeks**, we can side-step the issue with a new ``StringP`` type family "extension point" like the existing ``IdP`` one in **minutes**. [#extension-point]_

> Out of a basic fiduciary duty, we thus declare that unless "Plan A" works out almost as quickly, "Plan B" of just introducing another extension point should be used.

Reviewer: I don't think this is a fiduciary duty - those are duties incurred by virtue of being someone else's financial representative, where you must represent their interests. I'd say "To keep costs under control" or similar

Reviewer: Why does this need to be an extension point with a type family, rather than parameterising the AST over some arbitrary type of strings?

Reviewer 2: I think part of the point of TTG is that you just have _one_ parameter to the AST type, and all the other bits hang off that with type families. Otherwise you would have a truly ungodly number of type parameters.

Reviewer: Sounds good to me

The migration from |haskell-src-exts| to |ghc-lib-parser| was quite difficult because those libraries are nothing alike.
In contrast, we expect the migration from |ghc-lib-parser| to the new AST and parser libraries to be quite simple and pleasant, because the two new libraries should be very similar to |ghc-lib-parser|, and where they differ they should be strictly easier to use than before.

Note that HLint does use a few other things behind the AST and Parser that currently make it into |ghc-lib-parser|, but which we might want in our new libraries.

#. It uses GHC's multi-purpose ``Outputable`` instead of some more dedicated exact-printing machinary

#. It uses ``parseDynamicFilePragma``, and thus GHC's infamous ``DynFlags`` to support pragmas like ``{-# LANGUAGE ... #-}`` and ``{-# OPTIONS_GHC ... #-}``.

For the first case, we might consider factoring ``Outputable`` into a separate library too.
Or we can prioritize a more dedicated exact-print solution to use instead of ``Outputable`` (see the future work section).

> For the second case, we might have to do something temporary like e.g. continuing to use an auto-extracted library liek |ghc-lib-parser|, but depending on our newly factored-output libraries, to get this functionality for HLint_.

Reviewer: ```suggestion
For the second case, we might have to do something temporary like e.g. continuing to use an auto-extracted library like |ghc-lib-parser|, but depending on our newly factored-output libraries, to get this functionality for HLint_.
```
